sdk/queue/priority_queue.go

```diff
@@ -150,6 +150,46 @@ func (pq *PriorityQueue) PopByKey(key string) (*Item, error) {
 	return nil, nil
 }
 
+// GetByKey searches the queue for an item with the given key and returns it
+// if found.  The queue is unmodified.
+func (pq *PriorityQueue) GetByKey(key string) *Item {
+	pq.lock.Lock()
+	defer pq.lock.Unlock()
+
+	return pq.dataMap[key]
+}
+
+// Pop is used by heap.Interface to peak at the lowest priority item on the heap.
+func (pq *PriorityQueue) Peek() *Item {
+	pq.lock.RLock()
+	defer pq.lock.RUnlock()
+
+	i := pq.data.peek()
+	if i != nil {
+		return i.(*Item)
+	}
+	return nil
+}
+
+// WhileEmpty blocks the caller until the queue contains an item
+func (pq *PriorityQueue) WaitWhileEmpty(doneChan chan struct{}) {
+	for pq.Len() == 0 {
```

Comment:
This looks racy given there is no lock held. Once we enter the block length could be > 0